### PR TITLE
Functions and group by added to parser, updates to CMake structure

### PR DESCRIPTION
**Pull request by gaffneyk**
1. The parser is updated to support arbitrary functions, multiple selections, and the group by clause. It now produces a parse tree for queries of the form `SELECT Count(a), b FROM integer_table GROUP BY b;` and `SELECT CommonPrefix(ip) FROM ip_table;`.

2. I restructured the CMake files so each module links only with its downstream module, instead of linking all the libraries at the end. This allows us to add a dependency on any of the modules without worrying about downstream dependencies (see parser/test).

3. I re-enabled the parser tests and added the example queries above. After building the project, run the tests with `ctest --verbose`.

### CODE REVIEW

File: parser/test/parser_test.c [hunk 1]
```diff
@@ -1,21 +1,19 @@
...
-    parse_node *node = parse(test);
-
-    char *quickstep_output = quickstep_stringify(node);
-    printf("%s\n", quickstep_output);
-    free(quickstep_output);
+#include "parser/parser_wrapper.h"
+#include "parser/parse_node.h"
+#include "parser/utility/stringify.h"
 
+void test_parse_tree(char *query) {
```
Comment (johnchronis) [inline parser/test/parser_test.c]: Could you modify the test to verify the output of the parser against a known output, i.e. compare the json_output with a string containing the result you know to be correct.

File: parser/test/parser_test.c [hunk 2]
```diff
@@ -1,21 +1,96 @@
...
+                      "        \"from\": [\n"
+                      "          {\n"
+                      "            \"type\": \"ParseSimpleTableReference\",\n"
+                      "            \"table_name\": \"ip_table\"\n"
+                      "          }\n"
+                      "        ]\n"
+                      "      }\n"
+                      "    ]\n"
+                      "  }\n"
+                      "}";
```
Comment (gaffneyk) [inline parser/test/parser_test.c]: I know this is really ugly, but it's a temporary solution. We'll need to talk about testing frameworks.

File: parser/test/parser_test.c (hunk 1 above)
Comment (gaffneyk) [reply parser/test/parser_test.c]: Done!

File: parser/test/parser_test.c (hunk 2 above)
Comment (johnchronis) [reply parser/test/parser_test.c]: It is ok for now I think, but yes we should create a frameworks to test the output of the parser.
Also we should use testing frameworks like gtest

**Pull request merged**